feat: add linkedin auth

valentinyanakiev opened the pull request:
<!--

```
BREAKING CHANGES: This patch changes the behavior of configuration item `foo` to do bar. To keep the existing
behavior please do baz.
```
-->

Revives [original PR](https://github.com/ory/kratos/pull/2207) & replaces our (alstojanovic's) [previous PR](https://github.com/ory/kratos/pull/2792):

Rhetoric is that our organization has a commitment to a 3rd party for LinkedIn logins and our Identity is OSS, based on Ory Kratos. We are making a commitment on our side to contribute to Ory Kratos (with some help, hopefully) and get this social log-in inside the package.

The PR is draft as we'll need some support getting this in. There are a couple of weird issues with the flow:

1. Major blocking issue, [registering with one LinkedIn account and then trying to register and sign in with 2nd one ends up logging in with the first account. ](https://github.com/alkem-io/kratos/issues/1)
2. [Claims validation behaves strangely](https://github.com/alkem-io/kratos/issues/5), frequently the access token retrieved with the [3-legged LinkedIn flow](https://learn.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin%2Fcontext&tabs=HTTPS) returns that the token is revoked. 

Here is the flow:

 ![](https://i.imgur.com/peij7FB.png)


[The Kratos Login flow with LinkedIn has the following steps](https://learn.microsoft.com/en-us/linkedin/shared/authentication/authorization-code-flow?context=linkedin%2Fcontext&tabs=HTTPS):

1. User navigates to the Alkemio login page and choses 'Sign in with Linkedin'
2. First time Alkemio users coming from LinkedIn need to authorize Alkemio access their core information used in LinkedIn, namely user email (r_emailaddress) and basic (lite) profile info (r_liteprofile). After the user authorizes the application, linkedin calls back the redirectURI configured in the request
3. User receives an authorization token. User makes a request to replace it with Access Token.
4. The authorization token is replaced with Access Token. It is returned to the user, together with its expiry time.
5. Kratos LinkedIn provider calls the introspection endpoint with HTTP post. The request is form post, following redirects. NOTE: This has bugs as the application sometimes receives 200 but "active:false" [which means the credentials are valid but do not match the client information in the token](https://learn.microsoft.com/en-us/linkedin/shared/authentication/token-introspection?context=linkedin%2Fcontext&tabs=curl). This step is executed in the `Introspection` function of provider_linkedin.go in kratos.
6. We check the scopes in the Kratos config vs the authorized scopes, returned by the introspection call, `scope` claim. This step is executed in the `Claims` function of provider_linkedin.go in kratos.
7. We get the user profile. The profile provides the picture with [URN format](https://learn.microsoft.com/en-us/linkedin/shared/api-guide/concepts/urns?.context=linkedin%2Fcontext). We will need an additional call to LinkedIn API to get the full profile picture details. This step is executed in the `Claims` function of provider_linkedin.go in kratos - `err = l.ApiCall(ProfileUrl, &profile, exchange)`
8. We get the user email. This step is executed in the `Claims` function of provider_linkedin.go in kratos - `err = l.ApiCall(EmailUrl, &emailaddress, exchange)`

## Related issue(s)

https://github.com/ory/kratos/pull/2792
https://github.com/ory/kratos/pull/2207
<!--
If this pull request

1. is a fix for a known bug, link the issue where the bug was reported in the format of `#1234`;
4. is a fix for a previously unknown bug, explain the bug and how to reproduce it in this pull request;
5. implements a new feature, link the issue containing the design document in the format of `#1234`;
6. improves the documentation, no issue reference is required.

Pull requests introducing new features, which do not have a design document linked are more likely to be rejected and take on average 2-8 weeks longer to
get merged.

You can discuss changes with maintainers either in the Github Discussions in this repository or
join the [Ory Chat](https://www.ory.sh/chat).
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR.

Please be aware that pull requests must have all boxes ticked in order to be merged.

If you're unsure about any of them, don't hesitate to ask. We're here to help!
-->

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [x] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

aeneasr commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,179 @@
+	}
+	req.Header.Add("Authorization", bearer)
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
```
Please follow the implementations of other providers, specifically using

```
m.reg.HTTPClient(ctx, httpx.ResilientClientWithClient(o.Client(ctx, exchange)))
```

here. Example: 

https://github.com/ory/kratos/blob/2ff34b604757c46aae5cf3cbb23f39f982341486/selfservice/strategy/oidc/provider_facebook.go#L72

valentinyanakiev replied:
I have refactored the code using the resilient client and made it more readable and easy to follow

aeneasr commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+func (l *ProviderLinkedIn) Introspection(result interface{}, exchange *oauth2.Token) error {
+	resp, err := http.PostForm(string(IntrospectionURL),
```
Here too, please use the global HTTP client :)

valentinyanakiev replied:
Addressed, using the global http client everywhere

aeneasr reviewed (commented):
Thank you for the PR! I added some comments and suggestions.

From a high level, it looks like this is how the login should work. I can't say why LinkedIn returns that the token is not valid sometimes.

Unfortuantely the links to ZenHub are behind a login firewall, so I am not able to see the references.

valentinyanakiev commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,179 @@
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	err = json.Unmarshal(body, result)
```
Addressed + simplified where applicable

valentinyanakiev commented:
I have altered the claims response as it was mistakenly mapping `LocalizedFirstName` from the Linkedin payload to `name` in the `claims` object. Now using properly GivenName + Name. Tests are also updated to cover the change.

aeneasr commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,210 @@
+	}
+
+	claims := &Claims{
+		Email:     emailaddress.Elements[0].Handle.EmailAddress,
```
Both issuer and subject are missing, that's why the flow isn't working!

valentinyanakiev replied:
I couldn't find a standard `iss` claim for Linkedin, so I have, somewhat speculatively, defined it as `"https://login.linkedin.com/"`. Linkedin is not fully OIDC-compliant so some claims are not defined there. I am open to suggestions and feedback if you think something else would be more appropriate.

aeneasr commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,210 @@
+		Name:      fmt.Sprintf("%s %s", profile.LocalizedFirstName, profile.LocalizedLastName),
+		GivenName: profile.LocalizedFirstName,
+		LastName:  profile.LocalizedLastName,
+		Picture:   profile.ProfilePicture.DisplayImage.Elements[2].Identifiers[0].Identifier,
```
This smells like of out-of-bounds panics ;) Can we make this reliable?

valentinyanakiev replied:
Not only panic, we actually hit the bug with it. When the user doesn't have a picture this breaks the flow... Working on refreshing the PR with the suggestions.

valentinyanakiev replied:
I have put some checks around the code. I believe the critical case is actually when there is / isn't a profile picture - then the returned payload is different and quite simplified. I have added a negative scenario test where the picture is missing and added logic to fix it in the code. Tested it end-to-end, it should be working now.

aeneasr commented on selfservice/strategy/oidc/provider_linkedin.go:
```diff
@@ -0,0 +1,210 @@
+
+	claims := &Claims{
+		Email:     emailaddress.Elements[0].Handle.EmailAddress,
+		Name:      fmt.Sprintf("%s %s", profile.LocalizedFirstName, profile.LocalizedLastName),
```
We only want to include fields the upstream provides us - data changes we can do in the jsonnet itself:

```suggestion
```

techsmyth commented:
@nmapx yes, good news is that we got the linkedin flow working in a forked version of Kratos (https://github.com/alkem-io/kratos) just before christmas and have deployed it into production. It is on our hit list to polish up the PR with those fixes. We just simply ran out of time last year. Valentin is back from vacation on Monday and we can hopefully submit the updated PR next week. Any deadlines from Ory team that we should keep in mind?

nmapx commented:
@valentinyanakiev @techsmyth it has been 3 weeks already since last comment, any update here?

techsmyth commented:
@nmapx literally just chatted this through with Valentin. We have completed the merge of the 0.11.1 release into the local branch we used to test, all looks fine - so hopefully this week we can refresh this PR.

valentinyanakiev commented:
@aeneasr I have refreshed the PR, addressed comments + lint rules. Tests are passing, and the mocked responses are real payloads I have received (of course with altered IDs, PII, etc.) from LinkedIn. 

How do we take this further? I see documentation is in a separate repository, how do you normally proceed with that? Is it mandatory for this PR?

fboerman commented:
hi @valentinyanakiev  and @techsmyth I see this is merged in #3079 but I dont see any documentation on it. Could you perhaps share an example config for it to work? Thanks!
